### PR TITLE
Add contracts for Challenges

### PR DESCRIPTION
**Pull request by vinceau**
This PR adds contracts for the different ChaosDEX challenges.

### CODE REVIEW

File: ethereum-contracts/contracts/DEXReserve.sol [hunk 1]
```diff
@@ -53,13 +53,6 @@ contract DEXReserve is ERC20, ERC20Detailed, Ownable {
         }
     }
 
-    /// @notice Allow anyone to recover funds accidentally sent to the contract.
-    /// To withdraw ETH, the token should be set to `0x0`.
-    function recoverTokens(address _token) external {
```
Comment (noiach) [inline ethereum-contracts/contracts/DEXReserve.sol]: Was this throwing a compilation error?

File: ethereum-contracts/contracts/ChaosnetChallenges/Puzzle.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    /// @notice Allows someone to try and claim the reward by submitting the secret.
+    /// @param _refundAddress The address that should receive the shiftedOut tokens and the potential reward.
+    /// @param _secret The secret.
+    /// @param _amount The amount of Bitcoin provided to the Darknodes in Sats.
+    /// @param _nHash The hash of the nonce returned by the Darknodes.
+    /// @param _sig The signature returned by the Darknodes.
+    function claimReward(
+        // Payload
+        bytes memory _refundAddress,
```
Comment (loongy) [inline ethereum-contracts/contracts/ChaosnetChallenges/Puzzle.sol]: Using `_refundAddress` as the name is probably misleading. This is the address that will receive the claimed reward (there’s no “refund” happening).

File: ethereum-contracts/contracts/DEXReserve.sol (hunk 1 above)
Comment (noiach) [reply ethereum-contracts/contracts/DEXReserve.sol]: I see now - `function recoverTokens` appears twice.

File: ethereum-contracts/contracts/ChaosnetChallenges/Puzzle.sol (hunk 2 above)
Comment (vinceau) [reply ethereum-contracts/contracts/ChaosnetChallenges/Puzzle.sol]: But if the secret is incorrect then the funds that they shifted in is "refunded" back to them, no? Do you think `_rewardAddress` or something is more appropriate?

**Pull request merged**